quinn-proto/src/range_set.rs

```diff
@@ -8,6 +8,292 @@ use std::{
     },
 };
 
+use tinyvec::TinyVec;
+
+/// A set of u64 values optimized for long runs and random insert/delete/contains
+///
+/// `ArrayRangeSet` uses an array representation, where each array entry represents
+/// a range.
+#[derive(Debug, Default)]
+pub struct ArrayRangeSet(TinyVec<[Range<u64>; ARRAY_RANGE_SET_INLINE_CAPACITY]>);
+
+/// The capacity of elements directly stored in [`ArrayRangeSet`]
+///
+/// An inline capacity of 2 is chosen to keep `SentFrame` below 128 bytes.
+const ARRAY_RANGE_SET_INLINE_CAPACITY: usize = 2;
+
+/// An iterator over ranges in an [`ArrayRangeSet`]
+pub struct ArrayRangeSetIter<'a> {
+    inner: &'a TinyVec<[Range<u64>; ARRAY_RANGE_SET_INLINE_CAPACITY]>,
+    front_idx: usize,
+    back_idx: usize,
+}
+
+impl<'a> Iterator for ArrayRangeSetIter<'a> {
+    type Item = Range<u64>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.front_idx == self.back_idx {
+            None
```

Comment:
Style nit: in both the `Iterator` and `DoubleEndedIterator` impls, maybe just `return None;` eagerly in the simple case and remove the `else` in favor of continuing?